crates/core_simd/src/ops.rs

```diff
@@ -32,232 +31,497 @@ where
+/// At worst, this will maybe add another instruction and cycle,
+/// at best, it may open up more optimization opportunities,
+/// or simply be elided entirely, especially for SIMD ISAs which default to this.
+///
+// FIXME: Consider implementing this in cg_llvm instead?
+// cg_clif defaults to this, and scalar MIR shifts also default to wrapping
+macro_rules! wrap_bitshift {
+    ($(impl<const LANES: usize> $op:ident for Simd<$int:ty, LANES> {
+        fn $call:ident(self, rhs: Self) -> Self::Output {
+            unsafe { $simd_call:ident }
         }
-    } => {
-        impl<const $lanes: usize> core::ops::$trait<$rhs> for $type
+    })*) => {
+        $(impl<const LANES: usize> $op for Simd<$int, LANES>
         where
-            LaneCount<$lanes2>: SupportedLaneCount,
+            $int: SimdElement,
+            LaneCount<LANES>: SupportedLaneCount,
         {
-            type Output = $output;
+            type Output = Self;
 
-            $(#[$attrs])*
-            fn $fn($self_tok, $rhs_arg: $rhs_arg_ty) -> Self::Output $body
-        }
+            #[inline]
+            #[must_use = "operator returns a new vector without mutating the inputs"]
+            fn $call(self, rhs: Self) -> Self::Output {
+                unsafe {
+                    $crate::intrinsics::$simd_call(self, rhs.bitand(Simd::splat(<$int>::BITS as $int - 1)))
+                }
+            }
+        })*
     };
 }
 
-/// Automatically implements operators over vectors and scalars for a particular vector.
-macro_rules! impl_op {
-    { impl Add for $scalar:ty } => {
-        impl_op! { @binary $scalar, Add::add, simd_add }
-    };
-    { impl Sub for $scalar:ty } => {
-        impl_op! { @binary $scalar, Sub::sub, simd_sub }
-    };
-    { impl Mul for $scalar:ty } => {
-        impl_op! { @binary $scalar, Mul::mul, simd_mul }
-    };
-    { impl Div for $scalar:ty } => {
-        impl_op! { @binary $scalar, Div::div, simd_div }
-    };
-    { impl Rem for $scalar:ty } => {
-        impl_op! { @binary $scalar, Rem::rem, simd_rem }
-    };
-    { impl Shl for $scalar:ty } => {
-        impl_op! { @binary $scalar, Shl::shl, simd_shl }
-    };
-    { impl Shr for $scalar:ty } => {
-        impl_op! { @binary $scalar, Shr::shr, simd_shr }
-    };
-    { impl BitAnd for $scalar:ty } => {
-        impl_op! { @binary $scalar, BitAnd::bitand, simd_and }
-    };
-    { impl BitOr for $scalar:ty } => {
-        impl_op! { @binary $scalar, BitOr::bitor, simd_or }
-    };
-    { impl BitXor for $scalar:ty } => {
-        impl_op! { @binary $scalar, BitXor::bitxor, simd_xor }
-    };
+macro_rules! bitops {
+    ($(impl<const LANES: usize> BitOps for Simd<$int:ty, LANES> {
+        fn bitand(self, rhs: Self) -> Self::Output;
+        fn bitor(self, rhs: Self) -> Self::Output;
+        fn bitxor(self, rhs: Self) -> Self::Output;
+        fn shl(self, rhs: Self) -> Self::Output;
+        fn shr(self, rhs: Self) -> Self::Output;
```

Comment:
Personally I'm not a huge fan of so much of the pseudo-rust arguments, I actually find it harder to parse because it's not clear what's rust and what's not.  IMO now you have a whole line that's just a verbose way of saying "shr" or whatever.

Reply:
I should say that I would be fine with it if it weren't dramatically more verbose, but this doubles the length of the file with no significant functionality changes.

Reply:
A **few** extra lines actually helps me scan and do various text manipulation tricks, but I agree that I was aiming for... less than literally twice as many lines, yes. I will take another pass through it for savings.

Reply:
One more thing I found slightly misleading is the `impl BitOps` etc, because it makes me think there's an actual `BitOps` trait and not that it's shorthand for many

Reply:
Looking over this again, I think it's unnecessary to specify `BitOps` anyway, since it's also the name of the macro.  What about a single line like `impl BitAnd, BitOr, BitXor, Shl, Shr for i32` or whatever?

Reply:
Hm, the other way around, it should be like

`impl BitAnd for {i8, i16, i32, i64, isize}`
because the target types will remain consistent and the implementation for a _set_ of types (ints, uints, floats) will be consistent across that set, but a given implementation may vary subtly, and what we want to make easiest to edit is the implementation.